crates/ironrdp-displaycontrol/src/pdu/mod.rs

```diff
@@ -0,0 +1,628 @@
+//! Display Update Virtual Channel Extension PDUs  [MS-RDPEDISP][1] implementation.
+//!
+//! [1]: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpedisp/d2954508-f487-48bc-8731-39743e0854a9
+
+use ironrdp_pdu::cursor::{ReadCursor, WriteCursor};
+use ironrdp_pdu::{ensure_fixed_part_size, invalid_message_err, PduDecode, PduEncode, PduResult};
+
+const DISPLAYCONTROL_PDU_TYPE_CAPS: u32 = 0x00000005;
+const DISPLAYCONTROL_PDU_TYPE_MONITOR_LAYOUT: u32 = 0x00000002;
+
+const DISPLAYCONTROL_MONITOR_PRIMARY: u32 = 0x00000001;
+
+// Set out expectations about supported PDU values. 1024 monitors with 8k*8k pixel area is
+// already excessive, (this extension only supports displays up to 8k*8k) therefore we could safely
+// use those limits to detect ill-formed PDUs and set out invariants.
+const MAX_SUPPORTED_MONITORS: u16 = 1024;
+const MAX_MONITOR_AREA_FACTOR: u16 = 1024 * 16;
+
+/// Display Update Virtual Channel message (PDU prefixed with `DISPLAYCONTROL_HEADER`)
+///
+/// INVARIANTS: size of encoded inner PDU is always less than `u32::MAX - Self::FIXED_PART_SIZE`
+///     (See [`DisplayControlCapabilities`] & [`DisplayControlMonitorLayout`] invariants)
+#[derive(Debug, Clone)]
+pub enum DisplayControlPdu {
+    Caps(DisplayControlCapabilities),
+    MonitorLayout(DisplayControlMonitorLayout),
+}
+
+impl DisplayControlPdu {
+    const NAME: &'static str = "DISPLAYCONTROL_HEADER";
+    const FIXED_PART_SIZE: usize = 4 /* Type */ + 4 /* Length */;
+}
+
+impl PduEncode for DisplayControlPdu {
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        ensure_fixed_part_size!(in: dst);
+
+        let (kind, payload_length) = match self {
+            DisplayControlPdu::Caps(caps) => (DISPLAYCONTROL_PDU_TYPE_CAPS, caps.size()),
+            DisplayControlPdu::MonitorLayout(layout) => (DISPLAYCONTROL_PDU_TYPE_MONITOR_LAYOUT, layout.size()),
+        };
+
+        // This will never overflow as per invariants.
+        let pdu_size = payload_length.checked_add(Self::FIXED_PART_SIZE).unwrap();
```

Comment:
note: Alternatively, you can write using the normal operators and silent the clippy warning (justified by the comment).

```suggestion
        // This will never overflow as per invariants.
        #[allow(clippy::arithmetic_side_effects)]
        let pdu_size = payload_length + Self::FIXED_PART_SIZE;
```